### PR TITLE
Refactor RealmMetadata and remove TableHandle

### LINKED ISSUE #2213: Refactor Realm Metadata
At the moment `Realm.Metadata` is just a `Dictionary<string, RealmObjectBase.Metadata> Metadata`. It would be convenient to refactor it into a full fledged class, so it will be possible to add additional functionality and utility methods. In particular, we need to implement a method that allows to retrieve a `RealmObject`'s metadata from his `table_key`.

### PR DESCRIPTION
**Pull request by papafe**
This PR refactors `RealmMetadata` to create a correspondence between the `RealmObject.Metadata` and the `TableKey`. Given that with the `TableKey` we can get a reference to the table in native, `TableHandle` has been removed too.

Fixes #2213

### CODE REVIEW

File: Realm/Realm/Handles/ObjectHandle.cs
Comment (nirinchev) [inline Realm/Realm/Handles/ObjectHandle.cs:69]:
```suggestion
            public static extern IntPtr get_backlinks_for_type(ObjectHandle objectHandle, TableKey table_key, IntPtr source_property_index, out NativeException nativeException);
```

File: Realm/Realm/Handles/SharedRealmHandle.cs
Comment (nirinchev) [inline Realm/Realm/Handles/SharedRealmHandle.cs:104]: Try typedefing `realm_class_key` similar to what the C API does: https://github.com/realm/realm-core/pull/3955/files#diff-ac5ca02b88ba9910d618bac409ad7a1e3b281cdaf9f64840469c21870b7114a1R74

and have it as a return value.

File: Realm/Realm/Handles/SharedRealmHandle.cs
Comment (nirinchev) [inline Realm/Realm/Handles/SharedRealmHandle.cs:136]: snake_case for argument names

File: Realm/Realm/Native/TableKey.cs
Comment (nirinchev) [inline Realm/Realm/Native/TableKey.cs:29]: Not used - let's remove

File: Realm/Realm/Realm.cs
Comment (nirinchev) [inline Realm/Realm/Realm.cs:1367]: TODO: check where/why this is used.

File: Realm/Realm/Realm.cs
Comment (nirinchev) [inline Realm/Realm/Realm.cs:1389]: These should be expression-bodied

Comment (nirinchev) [reply Realm/Realm/Realm.cs:1389]: TODO: research the difference between expression bodied methods and regular ones.

File: Realm/Realm/Realm.cs
Comment (nirinchev) [inline Realm/Realm/Realm.cs:1399]: TODO: check if these are safely used.

File: wrappers/src/object_cs.cpp
Comment (nirinchev) [inline wrappers/src/object_cs.cpp:162]: maybe `const` here?

File: wrappers/src/shared_realm_cs.cpp
Comment (nirinchev) [inline wrappers/src/shared_realm_cs.cpp:255]: unneeded return here

File: wrappers/src/shared_realm_cs.cpp
Comment (nirinchev) [inline wrappers/src/shared_realm_cs.cpp:373]: Create an inlined helper method like `get_table(SharedRealm&, TableKey)`

File: wrappers/src/shared_realm_cs.cpp
Comment (nirinchev) [inline wrappers/src/shared_realm_cs.cpp:458]: Add `find(TableKey)` in OS's Schema class and use that instead.

File: wrappers/src/shared_realm_cs.cpp
Comment (nirinchev) [inline wrappers/src/shared_realm_cs.cpp:460]: Pass `ObjectStore::object_type_for_table_name(table->get_name())`

File: wrappers/src/shared_realm_cs.cpp
Comment (nirinchev) [inline wrappers/src/shared_realm_cs.cpp:487]: const

File: wrappers/src/table_cs.cpp
Comment (nirinchev) [inline wrappers/src/table_cs.cpp:54]: This file should be deleted.

File: Realm/Realm/Realm.cs
Comment (papafe) [reply Realm/Realm/Realm.cs:1367]: It is used in `realm.RemoveAll()`. It makes sense that it is used there, but probably would be more appropriate to transform the property in a method called `GetAllRealmObjectsMetadata()` or similar

Comment (nirinchev) [reply Realm/Realm/Realm.cs:1367]: It's probably best to just implement `RemoveAll` natively. It's obviously not a hot path, but it may have peculiar side effects - e.g. if you open the Realm file with a subset of its schema, you likely expect `RemoveAll` to leave the file empty, but it'll remove only some data.

You can either leave it as is and open an issue to refactor `RemoveAll` or go ahead and fix it in this PR 😛

Comment (papafe) [reply Realm/Realm/Realm.cs:1367]: I have added a new issue (#2233). Too much stuff already in this PR :)

File: wrappers/src/shared_realm_cs.cpp
Comment (papafe) [reply wrappers/src/shared_realm_cs.cpp]: I have just tried to implement this functionality as we discussed with @fealebenpae, by creating a map of `TableKey` to `ObjectSchema` when the `Schema` is created. Unfortunately the `table_key` field of `ObjectSchema` is not assigned yet in the constructor of `Schema`, and so we cannot create the map at that time. 
Given this, probably the fastest option would be to keep the code as it was as two operations are performed:

- `ObjectStore::object_type_for_table_name(table->get_name())` - O(1) time
- `realm->schema().find(object_name)` - O(log n) time (it uses binary search inside).

In alternative we can still implement find(TableKey) method, in two ways:

- We do a linear search - O(n) time
- We build the map the first time the method is called, and then use the map the following times - O(1) time + build time on first call

Comment (nirinchev) [reply wrappers/src/shared_realm_cs.cpp]: The big-O estimates for the string case appear to ignore the string operations themselves. Not a C++ expert, but I don't believe the language uses any sort of pooling, which means that constructing the string itself and comparing it is a O(m) operation, which is exacerbated by the fact all table names start with `class_`.

Perhaps ping @fealebenpae for suggestions as that's ultimately code his team will review and maintain 😄

Comment (papafe) [reply wrappers/src/shared_realm_cs.cpp]: As discussed with @nirinchev, we have decided to just use the linear search for now, then later we could eventually improve it

File: wrappers/src/object_cs.cpp
Comment (nirinchev) [inline wrappers/src/object_cs.cpp:154]: This can use the new `find(TableKey)` method.

File: wrappers/src/shared_realm_cs.cpp
Comment (nirinchev) [inline wrappers/src/shared_realm_cs.cpp:376]: `find(table_key)`

File: wrappers/src/sort_descriptor_cs.cpp
Comment (nirinchev) [inline wrappers/src/sort_descriptor_cs.cpp:45]: This can also use `find(TableKey)`.

**Review by nirinchev — COMMENTED** [review]
Mostly minor simplification suggestions from me, otherwise should be good to go once the Core PR is merged.

File: Realm/Realm/Native/TableKey.cs
Comment (nirinchev) [inline Realm/Realm/Native/TableKey.cs:31]:
```suggestion
        public override bool Equals(object obj) => obj is TableKey other && Equals(other);
```

**Review by nirinchev — APPROVED** [review]
Looks good but probably should be rebased on top of master to get the build fixed.

File: wrappers/src/shared_realm_cs.cpp
Comment (LaPeste) [inline wrappers/src/shared_realm_cs.cpp:382]: Are all these new lines needed?

File: wrappers/src/shared_realm_cs.cpp
Comment (LaPeste) [inline wrappers/src/shared_realm_cs.cpp:482]: Maybe this could be wrapped by curly brackets? I generally think it avoids issue.

**Review by LaPeste — CHANGES_REQUESTED** [review]
Beside the 2 minuscule comments I had, one more. ObjectKey is renamed to TableKey and you changed the year on the copyright. I'm not sure what the policy is but I assume all the other files you modified need to be changed, as well? If not, what's the policy about copyrights?

Anyway, beside the very small things mentioned above, PR looks good to me.

**Comment by papafe** [comment]
@LaPeste agree on both points. Regarding ObjectKey renamed to TableKey... it's just git that thinks that I renamed it. I actually deleted ObjectKey and added TableKey, but because they are similar git thinks I have renamed it, that's why the copyright difference.

**Comment by papafe** [comment]
This PR is considered mergeable, once the core changes have been approved (https://github.com/realm/realm-core/pull/4413) and merged to the v11 branch

**Review by LaPeste — APPROVED** [review]

**Comment by nirinchev** [comment]
The test failures are caused by a change in the Core query syntax. They'll be fixed once https://github.com/realm/realm-dotnet/pull/2214 is merged.

**Comment by nirinchev** [comment]
Failing tests seem to be my fault - the AOT dynamic runtime doesn't seem to like my code.

**Pull request merged**
